src/Reports/SalesByItems/Services/GroupingService.php

```diff
@@ -0,0 +1,230 @@
+<?php
+
+/**
+ * Service to group sales items and calculate subtotals
+ *
+ * Replaces the state-based logic from thisLineItem() in the original code.
+ * Groups items by category, then by product within each category.
+ * Calculates subtotals at both levels.
+ *
+ * @package   OpenEMR
+ * @link      https://www.open-emr.org
+ * @author    Sherwin Gaddis <sherwingaddis@gmail.com>
+ * @copyright Copyright (c) 2025 Sherwin Gaddis <sherwingaddis@gmail.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+namespace OpenEMR\Reports\SalesByItems\Services;
+
+use OpenEMR\Reports\SalesByItems\Model\SalesItem;
+use OpenEMR\Reports\SalesByItems\Model\SalesGroup;
+
+class GroupingService
+{
+    /**
+     * Group items by category and product
+     *
+     * Returns a structure like:
+     * [
+     *   'Office Visit' => [
+     *       '99213' => SalesGroup,
+     *       '99214' => SalesGroup,
+     *   ],
+     *   'Products' => [
+     *       'Bandage' => SalesGroup,
+     *   ]
+     * ]
+     *
+     * @param SalesItem[] $items The items to group
+     * @return array Grouped items with SalesGroup objects
+     */
+    public function groupItems(array $items): array
+    {
+        $grouped = [];
+
+        foreach ($items as $item) {
+            $category = $item->getCategory() ?: 'None';
+            $product = $item->getDescription() ?: 'Unknown';
+
+            if (!isset($grouped[$category])) {
+                $grouped[$category] = [];
+            }
+
+            if (!isset($grouped[$category][$product])) {
+                $grouped[$category][$product] = new SalesGroup($category, $product);
+            }
+
+            $grouped[$category][$product]->addItem($item);
+        }
+
+        return $grouped;
+    }
+
+    /**
+     * Get category totals (sum of all products in a category)
+     *
+     * @param SalesGroup[] $productGroups Array of SalesGroup objects in a category
+     * @return array ['quantity' => int, 'total' => float]
+     */
+    public function getCategoryTotals(array $productGroups): array
+    {
+        $quantity = 0;
+        $total = 0.0;
+
+        foreach ($productGroups as $group) {
+            $quantity += $group->getQuantity();
```

Comment:
The SalesGroup has a facility for adding up quantities and totals, but that work is repeated here.